packages/shared/src/introspectMachine.ts

```diff
@@ -7,6 +7,47 @@ export interface SubState {
   states: Record<string, SubState>;
 }
 
+function getRelevantFinalStates(node: XState.StateNode): XState.StateNode[] {
+  switch (node.type) {
+    case "compound":
+      return getChildren(node)
+        .map((childNode) =>
+          childNode.type === "final"
+            ? [childNode]
+            : getRelevantFinalStates(childNode)
+        )
+        .flat();
+    case "parallel":
+      const finalStatesPerRegion = getChildren(node).map(
+        getRelevantFinalStates
+      );
+      return finalStatesPerRegion.every((perRegion) => !!perRegion.length)
+        ? finalStatesPerRegion.flat()
+        : [];
+    default:
+      return [];
+  }
+}
+
+function getAllNodesToNodes(nodes: XState.StateNode[]) {
```

Comment:
I'm confused, what's the purpose of this function?

Reply:
The name could definitely be improved - i didn't figure out a better one for this. This function grabs all the nodes from a set of nodes to the root.

<img width="652" alt="Screenshot 2022-07-28 at 20 55 56" src="https://user-images.githubusercontent.com/9800850/181616114-ae6f7d1b-044e-40cf-bfde-1a50b411f6f3.png">

On this graph, the red circles are the input set and all the green ones are part of the result.

In practice, I don't even need this result more than once. I just end up iterating over it to add the events to all of those states and I could just do it while iterating till the root. I've extracted this to a separate function to split those two concerns here and to make it easier to read but perhaps this wasn't really necessary.

Reply:
Ah! This is the `getConfiguration(...)` function in v5.

Reply:
Hm, it's close to that but it's not exactly the same and perhaps my picture was a little bit misleading here. The difference is that a valid configuration always contains states up to the leaves but this accepts any set of nodes as inputs. In practice, I call it with a set of parallel states (so kinda roots of those "leaf parallel" regions).

Reply:
in a machine like this:
```scxml
<scxml>
  <parallel>
    <parallel>
      <state/>
```

I call the inner parallel a "leaf parallel" but its parent parallel state (or any above it) is not considered as such).